---
title: "Transmission_manuscript"
author: "Karim MANE"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  word_document: default
  html_document: 
    df_print: paged
  pdf_document: default
bibliography: references.bib
link-citations: true
csl: biomed-central.csl
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
suppressPackageStartupMessages(library("dplyr"))
suppressPackageStartupMessages(library("data.table"))
suppressPackageStartupMessages(library("mpbr"))
suppressPackageStartupMessages(library("ggplot2"))
suppressPackageStartupMessages(library("ggpubr"))
suppressPackageStartupMessages(library("rstatix"))
suppressPackageStartupMessages(library("tictoc"))

# DEFINITION OF THE INPUT FILES
path_to_data <- "/Users/karimmane/Documents/Karim/Personnel/Transmission_paper/Data"
vcf <- file.path(path_to_data, "Input_Data.vcf.gz")
metadata <- file.path(path_to_data, "SampleMetadata.txt")
out_dir <- file.path(path_to_data, "Processed_data")

# DEFINITION OF SOME CONSTANTS
maf_cutoff <- 0.01
missingness_cutoff <- 0.2
```

# PROPOSED TITLE
Persistent _Plasmodium falciparum_ genome-wide gene-flow and malaria transmission within The Gambia and across the border with Senegal  

# ABSTRACT
The availability of high throughput sequencing data has made it possible to explore individual genomes to understand the basis of population evolution. Combined with the development and application of statistical and population genetics approaches it's now possible to investigate individual relatedness  between and/or within populations. Among other metrics, **IBD (identity by descent)** estimates has become one of the commonly used method to quantify relatedness and divergence at both individual and population level. The outputs from such approaches can be used to investigate the evolutionary dynamics of pathogens. In this study, it was used to track the malaria parasite, _P. falciparum_ population movement and identify malaria reservoir in the Senegambia area. We used a dataset consisting of 16,390 Single Nucleotide Polymorphisms (SNPs) from whole genome sequencing of 875 _P. falciparum_ isolates collected from 3 Senegalese and 10 Gambian sites between 1984 and 2015. The results showed the existence of gene flow between Senegal and The Gambia, but at a lower extent than within The Gambia only.    

**NB1:** need to expand more on the method, results and conclusion in the abstract.   

**NB2:** relatedness was estimated after removal of drug resistance genes. Without these highly mutant regions of the genome, there is likely that old isolate and recent one to be similar, except if there are other mutation forces driving the evolution of the parasites.     

**NB3:** to check NB2, relatedness needs to be estimated both with and without drug resistant genes.     

**Question1:** the observed relatedness results are they due to human mobility or due to other factors (use of different policies across different regions, similar evolution due to similar environment, etc)?

# INTRODUCTION
The risk of malaria is now generally low in most of The Gambia and Senegal following a decade of scaled-up interventions with long-lasting insecticide treated nets, the use of artemisinin-based combination therapies (ACTs) and more recently, the implementation of seasonal malaria chemoprevention (SMC). The Gambia and Senegal are now part of 15 countries in West Africa and the Sahel vying for malaria elimination[@Proof1]. However, malaria transmission across both countries remains very heterogeneous, with higher intensity around the southern and eastern regions of Senegal and relatively low transmission northwards towards Mauritania[@Proof2]. The Gambia is completely surrounded by Senegal except for the Atlantic coast and transmission is highly seasonal, spanning the rainy season from July to November with a peak in October[@Proof3]. Transmission is lower (less than <5%  infection prevalence) in the coastal western regions of the country but higher eastwards where prevalence range from 10-40%[@J_Mwesigwa]. This difference can be due to many reasons, including differences in the density and species of mosquitoes as well as the more rural nature of human settlements with poorer housing in the eastern regions. Parasite migration, through human mobility, from high to low prevalence areas is an important obstacle in the malaria interruption. With the increasing levels of human migration within The Gambia and across the border with Senegal, elimination approaches can be refined by targeting hotspots and breaking parasite flow between hotspots and cold spots of infection. Hence, the National Malaria Control programs of Senegal and The Gambia are in collaboration to synchronize approaches and timing of interventions. To support this effort there is need determine changes in the parasite populations imposed by interventions and importantly the migration patterns linking temporal and geographic populations. This transmission dynamics can be refined by analyzing the relatedness between parasite isolates with defined epidemiological information.     

Several studies have shown the importance of _P. falciparum_ genetic data in resolving connectivity between parasite populations[@S_K_Tessema_2019, @R_Noviyanti_2020]. Earlier studies evolved from the use of the microsatellites to the 24 SNPs barcode. The latter have been used in Senegal to describe changes in parasite populations following years of interventions[@Proof4]. However, with the advent of Next Generation Sequencing (NGS) and high-density genotyping platforms, extended SNP barcodes and whole genome variation data have replaced the 24 SNPs which is limited in resolving highly related populations. The combination of the NGS data and the use of statistical approaches, that allow to increase the accuracy of determining relatedness, is helping in investigate the extend of local and imported pathogens in a population. Recently utilized in Southeast Asia, the approaches revealed the potential origins of  migrating parasites[@A_C_Shetty_2019], and those that carry drug resistance associated mutations. In Africa, _P. falciparum_ genetic data was combined with travel history to show that the majority of the malaria cases in the northeastern Namibia was due to local transmission[@S_K_Tessema_2019_elife], while in The Gambia, this permitted to detect several transmission paths across the country[@A_Amambua-Ngwa_2019]. Beyond these general transmission questions, defining the co-ancestries of isolates with potential for drug resistance and their dispersal is highly relevant, given the importance of antimalarial drugs in combating morbidity and mortality from malaria. Also, genetic relatedness can help identify outbreaks of epidemic malaria as demonstrated in Tanzania where clusters of higher malaria incidence were found in a north-eastern district[@Teun_Bousema_2010].

Several statistical and population genetics approaches that were used to measure the degree of similarity between isolates of the same pathogen were mostly based on distance matrices and clustering approaches[@Proof5], which recently have been argued to be less accurate in defining the degree and timing of shared ancestry and therefore transmission links between individual isolates separated by space or time[@Proof6]. Thus, the relationship between the expected length of haplotype sharing between isolates and the time since divergence from a common ancestor has been widely used in reconstructing the genealogical history of human populations[@P_Ralph_2013]. Among other metrics, the identity-by-state (IBS) and the identity-by-descend (IBD) have been proposed as better matrices for ancestry for malaria parasites, as they consider the added complications of selfing that could occur during sexual recombination in the mosquito vector. The IBD approach is an increasingly used method of inferring recent demographic event and connectivity between populations [@A_Amambua-Ngwa_2019]. With IBD, shorter haplotype sharing is suggesting of distant ancestry and longer haplotype sharing reflects more recent links between isolates. As IBD can therefore define clusters of infection, epidemic outbreaks and geneflow between temporal and spatial populations, approaches to transform IBD matrices into information for tracking parasite movements in time and space should be able to distinguish the most current patterns in order to obtain informative insights that can be used by the NMCPs. Here, we used a previously published method[@S_K_Tessema_2019_elife] to infer genome wide IBD from SNPs data to show the extend of gene flow and identify malaria reservoirs in the Senegambia regions.


## MATERIALS AND METHODS

### Quality filtering of the SNPs data

The data consisted of Variant Call File (VCF) of isolates collected from 11 sites (2 and 9 in Senegal and The Gambia respectively), between the year 1984 and 2015, from several projects in collaboration with the malariaGEN[@MalariaGEN_ressources] programme at the Wellcome Sanger institute (1103-PF-PDN-GMSN-NGWA, 1136-PF-GM-NGWA, 1137-PF-GM-DALESSANDRO) and freely available data of isolates from Senegal obtained via the WSI Pf3K project. Apart from Farafenni and Serrekunda in The Gambia, where samples were collected between 1984 and 2000, the Gambian isolates were sampled recently (in 2008 and between 2011 and 2015). Isolates were sequenced on Illumina platforms[@Illumina] and the GATK pipeline was used to generate the VCF files as earlier described[@GATK]. Biallelic Single Nucleotide Polymorphisms (SNPs) were extracted from the VCF files and filtered with the following criteria: Minor allele frequency (MAF) > 1%, RMS mapping quality (MQ) > 20, read depth (DP) > 5 and VQSLOD > 3. 

```{bash "vcf_filtration", eval=FALSE, echo=FALSE}
# add the bcftools command for filtering the input VCF file. Look at the VCF file header

# EXTRACT SNPS WITH MAF>=0.01 & MQ > 20 && DP > 5 && VQSLOD > 3
bcftools view -q 0.01:minor -i'MQ > 20 && DP > 5 && VQSLOD > 3'-o Input_Data.vcf.gz -O z original.vcf.gz
```

SNPs passing this initial filtration criteria were further filtered based on the percentage of missing genotypes to retain only loci with less than 20% of missing data across all samples. Isolates were also discarded if they had more than 20% missing calls for SNP loci. 

Note that the R functions used in this analysis are now part of the R package[@rSNPdata].

```{r "create_snpdata_object", eval=TRUE, echo=FALSE}
# add the R code to used to filter SNPs and samples based on missingness

# CREATE THE DIRECTORY THAT WILL CONTAIN THE TEMPORARY FILES
system(sprintf("mkdir -p %s", out_dir))

# CREATE A `snpData` OBJECT REQUIRED TO USED THE FUNCTIONS IN rSNPdata
snpdata <- rSNPdata::get_snpdata(
  vcf.file = vcf, 
  meta.file = metadata, 
  output.dir = out_dir, 
  gaf = NULL, 
  gff = NULL 
)

#In the above command used to create the `snpData` object, we are using the default gene annotation and ontology files that are stored in the `rSNPdata` package.
```

### Initial QC

The minor allele was determined by considering only the reference and alternate alleles. We ignored the mixed genotypes at this stage given they will be phased and transformed into either the reference or alternate allele.

```{r "initial_qc", eval=TRUE, echo=FALSE}
# CALCULATE AND DISPLAY THE MAF
snpdata <- rSNPdata::compute_MAF(
  snpdata, 
  include.het=FALSE, 
  mat.name="GT"
)

# GENERATE THE DATA FRAMES TO USE WHEN DISPLAYING THESE METRICS  
# IN THE RESULTS SECTION
maf <- data.frame(cbind(
  MAF = snpdata$details$MAF,
  Filtration = "before"))

snps_missingness <- data.frame(cbind(
  missingness = snpdata$details$percentage.missing.samples,
  Filtration = "before"))

samples_missingness <- data.frame(cbind(
  missingness = snpdata$meta$percentage.missing.sites,
  Filtration = "before"
))
```

### Data filtration

Keeping only sites and samples that satisfy the defined conditions in the above section. 

```{r "data_filtration", eval=TRUE, echo=FALSE}
# PERFORM QC
snpdata <- rSNPdata::filter_snps_samples(
  snpdata, 
  min.qual = 10, 
  max.missing.sites = 0.2, 
  max.missing.samples = 0.2, 
  maf.cutoff = 0.01
)

# COMPUTE THE MAF
snpdata <- rSNPdata::compute_MAF(
  snpdata, 
  include.het = FALSE, 
  mat.name = "GT"
)

# GET THE METRICS
tmp_maf <- data.frame(cbind(
  MAF = snpdata$details$MAF,
  Filtration = "after"))
tmp_maf$MAF <- as.numeric(tmp_maf$MAF)

# CALCULATE THE SAMPLE AND SNPs MISSINGNESS AFTER FILTRATION
tmp_samples_missingness <- data.frame(cbind(
  missingness = colSums(is.na(snpdata$GT)) / nrow(snpdata$GT),
  Filtration = "after"
))

tmp_snp_missingness <- data.frame(cbind(
  missingness = rowSums(is.na(snpdata$GT)) / ncol(snpdata$GT),
  Filtration = "after"
))

# UPDATE THE SAMPLE AND SNP MISSINGNESS
snpdata$meta$percentage.missing.sites <- tmp_samples_missingness$missingness
snpdata$details$percentage.missing.samples <- tmp_snp_missingness$missingness


# BINDING THE METRICS COLLECTED BEFORE AND AFTER
maf <- rbind(maf, tmp_maf)
maf$Filtration <- factor(maf$Filtration, levels = c("before", "after"))
maf$MAF <- as.numeric(maf$MAF)
samples_missingness <- rbind(samples_missingness, tmp_samples_missingness)
samples_missingness$Filtration <- factor(samples_missingness$Filtration, 
                                         levels = c("before", "after"))
samples_missingness$missingness <- as.numeric(samples_missingness$missingness)
snps_missingness <- rbind(snps_missingness, tmp_snp_missingness)
snps_missingness$Filtration <- factor(snps_missingness$Filtration, 
                                         levels = c("before", "after"))
snps_missingness$missingness <- as.numeric(snps_missingness$missingness)

# PLOTING THE RESULT AND USE IT IN THE RESULTS SECTION
sample_missingness_plot <- ggplot2::ggplot(
  samples_missingness, aes(x = missingness)) +
  geom_histogram() +
  labs(title = "sample missingness distribution", x = "missingness", 
       y = "Count") +
  theme_bw() +
  theme(legend.position = "none") +
  facet_grid(~Filtration)

maf_plot <- ggplot2::ggplot(maf,aes(x = MAF)) + 
  geom_histogram()+
  labs(title = "MAF distribution", x = "MAF", y = "Count") +
  theme_bw() +
  theme( legend.position = "none") +
  facet_wrap(~Filtration) 

snp_missingness_plot <- ggplot2::ggplot(snps_missingness,aes(x = missingness)) + 
  geom_histogram()+
  labs(title = "SNPs missingess distribution", x = "missingness", y = "Count") +
  theme_bw() +
  theme( legend.position = "none") +
  facet_wrap(~Filtration) 
```

### Population diversity 

We estimated the complexity of infection (COI) by computing the within-hosts fixation index (Fws). This metric measures the genetic diversity of an individual infection relative to the population level genetic diversity[@Fws] using the moimix R package[@moimix]. We also compared the within host diversity distribution between the different sampling regions through a Wilcoxon test using the rstatix[@rstatix] R package and the obtained p-values were adjusted based on the Bonferroni method. The 4 geographical regions are defined as follow: **western region** (Serrekunda, Brikama, Besse), **central region** (Chogen, Farafenni, Dongoro Ba), **eastern region** (Basse Njayel, Sare Wuro), **Senegal** (Thies, Bounkiling).

```{r "fws", eval=TRUE, echo=FALSE, include=FALSE}
# CALCULATE Fws
snpdata <- calculate_Fws(snpdata)
snpdata$meta$Fws <- as.numeric(snpdata$meta$Fws)

snpdata$meta$Location <- factor(snpdata$meta$Location, 
                                levels = c("Serrekunda", "Brikama", "Besse",
                                           "Chogen", "Farafenni", "DongoroBa",
                                           "Basse", "SareWuro", "Njayel",
                                           "Thies", "Bounkiling")
                                   )
per_location_fws <- ggplot(
  snpdata$meta, aes(x=Location, y=Fws, fill=Location)) + 
  geom_boxplot(outlier.size = 0.1) +
  scale_fill_manual(values = rep('gray', 
                                 length(levels(snpdata$meta$Location)))) +
  theme(
      plot.background = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.border = element_blank(),
      panel.background = element_blank(),
      axis.text.x = element_text(face="plain", color="black", angle = 60, 
                                 hjust = 1, size=6),
      axis.text.y = element_text(face="plain", color="black", size=6),
      axis.line = element_line(colour = "black", size = 0.5, 
                               linetype = "solid"),
      legend.position = "none"
  )

# COMPARING MEAN FWS BETWEEN REGIONS
snpdata$meta$regions <- "Senegal"
snpdata$meta[which(snpdata$meta$Location == 'Brikama' |
                     snpdata$meta$Location == 'Besse') , ]$regions <- 'Western'
snpdata$meta[which(snpdata$meta$Location == 'Basse' |
                  snpdata$meta$Location == 'Njayel' |
                  snpdata$meta$Location == 'SareWuro') , ]$regions <- 'Eastern'
snpdata$meta[which(snpdata$meta$Location == 'Serrekunda' |
                  snpdata$meta$Location == 'Farafenni') , ]$regions <- 'Ancient'
snpdata$meta[which(snpdata$meta$Location == 'DongoroBa' |
                     snpdata$meta$Location == 'Chogen') , ]$regions <- 'Central'
snpdata$meta$regions <- factor(snpdata$meta$regions, 
                               levels = c("Western", "Central", "Eastern",
                                          "Ancient", "Senegal")
                               )
stat.test <- snpdata$meta %>% 
  wilcox_test(Fws ~ regions, p.adjust.method = "bonferroni") # note that we will consider the adjusted p-values 
saveRDS(stat.test, file.path(out_dir, "Fws_Stat_ByRegion.RDS"))
per_region_fws = ggpubr::ggboxplot(snpdata$meta, x="regions", y="Fws", fill = "regions",
                palette = c("#00AFBB", "black", "#FC4E07", "#E7B800","gray"),
                ylab = "Fws", xlab = "regions") +
    ggpubr::stat_pvalue_manual(
        stat.test, label = "p.adj.signif", tip.length = 0.01,
        y.position = c(1.1, 1.2, 1.5,1.6,1.1,1.3,1.4,1.1,1.2,1.1),
        bracket.shorten = 0.05
    ) +
  theme(
    axis.text.x = element_text(face="plain", color="black", angle = 60,
                               hjust = 1, size=6), #, size=14
    axis.text.y = element_text(face="plain", color="black", size=6),
    legend.position = "none",
    axis.line = element_line(color = 'black', size = 0.5, linetype = 'solid')
  )
```


### Handling drug resistance regions and mixed genotypes

Many regions of the genome of the malaria parasite _P. falciparum_ are part of genes that are associated with antimalarial drugs resistance. They include chloroquine resistance transporter (Pfcrt) on chromosome 7, dihydrofolate reductase (Pfdhfr) on chromosome 4, dihydropteroate synthetase (Pfdhps) on chromosome 8, the multidrug resistance gene 1 (Pfmdr1) on chromosome 5, Pfaat1 on chromosome 6, Pfnfs on chromosome 7, Pfsmp1 and Pfmsp2 on chromosome 10, the one on chromosome 9, Pfk13 on chromosome 13 and the other one on chromosome 13.   

```{r "drug_resistant_regions", echo=FALSE, eval=TRUE, warning=FALSE}
resistance_regions <- data.table::fread(
  file.path(path_to_data, "Resistance_Gene.txt")
  )
resistance_regions$gene <- c("Pfdhfr", "Pfmdr1", "Pfaat1", "Pfnfs", "Pfcrt", "Pfdhps", NA_character_, "Pfmsp1", "Pfmsp2", "Pfk13", NA_character_)
knitr::kable(resistance_regions, format = " pipe", align = 'c', 
             caption = "table1: regions of the _P. falciparum_ genome associated with drug resistance") %>% 
  kableExtra::kable_styling("striped", position = "center") %>% 
  kableExtra::scroll_box(width = "95%")
```

As these regions are generally well conserved[@Proof7] in the different isolates, we discarded them to reduce their potential influence on the inferred relatedness values.
Similarly, as mixed genome infections could confound the determination of the population structure and relatedness between infections, we proposed the following 3 approaches of replacing mixed genotypes:    
a) **raw** where the mixed alleles are considered as a third allele and recoded as `2`, b) **minor_David** where mixed alleles are replaced based on the minor allele at the given SNP position, and c) **minor_Karim** where the mixed alleles are recoded based on the minor allele obtained from the number of reads supporting each allele at the given position for the given isolates. 
In both approaches, the reference allele remained coded as `0`, and the alternate as `1`.

### Phasing the mixed genotypes

We have implemented `minor_David` as our genotype phasing method in `rSNPdata`. This approach was shown to be the most appropriated way of phasing the mixed genotypes. 

```{r "genotype_phasing", eval=TRUE, echo=FALSE}
# PHASING THE MIXED GENOTYPES USING 15 ITERATIONS (CONSIDER 100+ ITERATIONS FOR MORE ACCURACY)
tictoc::tic()
snpdata = rSNPdata::phase_mixed_genotypes(
  snpdata, 
  nsim=15
)
tictoc::toc()
```

### Imputing the missing genotypes

To infer relatedness based on Aimee Taylor's method, the input genotype matrix should not contain missing genotypes. We implemented an imputation method based on the **MAF calculated from the allelic depth** at every missing locus for every sample.  

```{r "genotype_imputation", eval=TRUE, echo=FALSE}
# IMPUTING THE MISSING GENOTYPES FROM THE PHASED DATA USING 15 ITERATIONS (CONSIDER 100+ ITERATIONS FOR MORE ACCURACY)
snpdata = rSNPdata::impute_missing_genotypes(
  snpdata, 
  genotype="Phased", 
  nsim=15
)
```

### Relatedness between _P. falciparum_ isolates	

**This section needs to be implemented in the method section. Also the result of the comparison of the 3 methods need to be shown to justify the implementation of David's method.**

#### Determination of the optimal MAF for relatedness inferrence

We estimated relatedness between pairs of isolates from 2 distinct populations or the same population based on a model proposed by Aimee Taylor et and coauthors[@Aimee_R_Taylor_IBD]. Between every pairs of population, we generated 5 different data sets corresponding to a specific minor allele frequency (MAF) cut-off (MAF>=1%, MAF>10%, MAF>20%, MAF>30%, MAF>40%). 

#### Determination of the optimal mixed genotypes phasing approach for relatedness inferrence

For each MAF cut-off, we estimated relatedness based on the above method. To determine the optimal MAF cut-off for the rest of the analysis, we studied the variation of the mean relatedness from every MAF cut-off. 
Finally, the 3 ways of phasing the mixed genotypes were compared by calculating, for each MAF cut-off, the correlation coefficients between the relatedness values obtained by considering the mixed allele as a 3rd allele and the 2 other methods.

```{r eval=TRUE, echo=FALSE, include=FALSE}
# ADD SCRIPT
```


### Detecting highly related infection pairs 
The estimated relatedness between pairs of isolates range from 0 (when the 2 isolates are unrelated) to 1 (when the 2 isolates are clonal). We considered any pair of infections with a relatedness value greater than or equal to 0.6 as a pair of highly related infections (HRI) and calculated the proportion of HRI pairs between and within the 11 populations. The result is shown on a map where populations are connected by a line which width is proportional to the inferred relatedness. The GPS coordinates of the lines that were used to connect 2 sites were generated using a customized python script adapted from[@Line_in_map].   

```{r eval=TRUE, echo=FALSE, include=FALSE}
# ADD SCRIPT
```


### Malaria transmission dynamic between high parasite prevalence villages
Malaria transmission is still high in the east of the Gambia. The pattern of the transmission dynamic within and between 4 villages around that region was studied. The largest distance between these locations is about 30 Kilometers. A total of 145 samples from these 4 villages had complete GPS coordinates. 
In this section, we considered the 183 pairs of infections with at least a relatedness value greater than or equal to 0.2 from these 145 samples and compared the relatedness distribution between pairs of isolates from the same village against the one from different villages. On the other hand, the association between the relatedness and the distance between the sampling sites was assessed using the quantreg R package[@quantreg]. Finally, we checked whether the genomic sharing is higher in pairs of infections collected in the same compound than in those from different compounds and established if the difference can be associated or not to the distance between the 2 compounds. 

```{r eval=TRUE, echo=FALSE, include=FALSE}
# ADD SCRIPT
```

## REFERENCES





