---
title: "Population_genetics_analysis"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Population_genetics_analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(collapse = TRUE, comment = "#>")
vcf        <- system.file("extdata", "Input_Data.vcf.gz", package = "mpbr")
meta_file  <- system.file("extdata", "SampleMetadata.RDS", package = "mpbr")
output_dir <- tempdir()
```

```{r setup}
library(mpbr)
```

## Fst

In this package, Weir & Cockerham's Fst is calculated using the `vcflib` tools.   
The `from` option specifies the name of the column in the metadata table with the groups between which Fst should be calculated.

The `calculate_wcFst` returns a `SNPdata` object with an additional field named `Fst`. It is a `list` that contains data frames representing each the Fst results between a given pair-wise population. 

```{r eval=FALSE}
# CREATE THE SNPdata OBJECT
snpdata <- get_snpdata(vcf_file   = vcf,
                       meta_file  = meta_file,
                       output_dir = output_dir)

# COMPUTE THE SNPs MINOR ALLELE FREQUENCIES
snpdata <- compute_maf(snpdata,
                       include_het = FALSE,
                       mat_name    = "GT")

# FILTER SNPs AND SAMPLES
snpdata <- filter_snps_samples(snpdata,
                               min_qual            = 10,
                               max_missing_sites   = 0.2,
                               max_missing_samples = 0.2,
                               maf_cutoff          = 0.01)

# CALCULATE THE WITHIN HOST FIXATION INDEX Fws
snpdata <- calculate_fws(snpdata, threshold = 0.95)

# CALCULATE Fst BETWEEN SENEGAL AND GAMBIA
tic()
snpdata <- calculate_wcFst(snpdata,
                           from = "Country",
                           groups = c("Senegal", "Gambia"))
toc()
names(snpdata$Fst)
head(snpdata$Fst$Senegal_vs_Gambia)
```

## IBS 

The identity by state (IBS) represents the proportion of similar loci between 2 pairs of isolates. The `calculate_IBS` function returns the dissimilarity matrix i.e. (1-IBS). It returns a `SNPdata` object with an additionnal field named `IBS`. 

```{r eval=FALSE}
snpdata <- calculate_IBS(snpdata, mat.name = "GT")
```

## LD   

The linkage between pairs of SNPs is calculated using the `vcftools` program.

```{r eval=FALSE}
## calculate LD between pairs of SNPs across chromosome 4 and 5
snpdata <- calculate_LD(snpdata, min.r2 = 0.2, inter.chrom = FALSE,
                        chroms = c("Pf3D7_04_v3", "Pf3D7_05_v3"))

## calculate LD between pairs of SNPs across all chromosomes
snpdata <- calculate_LD(snpdata, min.r2 = 0.2, inter.chrom = FALSE,
                        chroms = NULL)
```

## iR

```{r eval=FALSE}
groups <- c("DongoroBa", "Chogen")
idx <- which(snpdata$meta$Location %in% groups)
samples.to.be.dropped <- snpdata$meta$sample[-idx]
reduced_snpdata <- drop_samples(snpdata, samples.to.be.dropped)
print(reduced_snpdata)
res.iR <- calculate_iR(reduced_snpdata, mat.name = "Phased",
                       family = "Location", number.cores = 4)
dim(res.iR$iR$Chogen_vs_DongoroBa)
```

## relatedness

relatedness between pairs of isolates from a given pairs of population. The relatedness is calculated using the `GT` table. The value of the `from` option is the name of the column in the metadata data frame that represents the sample's population of origin. To remove regions of the genome under selection, use the `sweepRegions` option. You can specify a vector of populations to compute the relatedness for as a value for the groups option.  

```{r eval=FALSE}
## calculate relatedness between pairs of isolates from Chogen and DongoroBa
snpdata <- impute_missing_genotypes(snpdata, genotype = "GT", nsim = 10)
snpdata <- calculate_relatedness(snpdata, mat.name = "Imputed",
                                 from = "Location", sweepRegions = NULL,
                                 groups = c("Chogen", "DongoroBa"))
relatedness <- snpdata$relatedness$df
relatedness.matrix <- snpdata$relatedness$matrix

## calculate relatedness between pairs of isolates from all locations
snpdata <- calculate_relatedness(snpdata, mat.name = "GT", family = "Location",
                                 sweepRegions = NULL, groups = NULL)

## calculate relatedness between pairs of isolates from all locations after
## excluding the selective sweep regions
Chrom <- c("Pf3D7_04_v3", "Pf3D7_05_v3", "Pf3D7_07_v3")
Start <- c(723088, 932890, 378222)
End <- c(774914, 987149, 431317)
selection.region <- data.frame(cbind(Chrom, Start, End),
                               stringsAsFactors = FALSE)
snpdata <- calculate_relatedness(snpdata, mat.name = "GT", family = "Location",
                                 sweepRegions = selection.region, groups = NULL)
```
