<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Create the SNPdata object — get_snpdata • mpbr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Create the SNPdata object — get_snpdata"><meta property="og:description" content="This function generate the input data needed for whole genome SNPs data
genotyped from malaria parasite."><meta property="og:image" content="https://Karim-Mane.github.io/mpbr/logo.svg"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mpbr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.1</span>
      </span>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/mpbr.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/Karim-Mane/mpbr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Create the <code>SNPdata</code> object</h1>
    <small class="dont-index">Source: <a href="https://github.com/Karim-Mane/mpbr/blob/HEAD/R/get_snpdata.R" class="external-link"><code>R/get_snpdata.R</code></a></small>
    <div class="hidden name"><code>get_snpdata.Rd</code></div>
    </div>
    <div class="ref-description">
    <p>This function generate the input data needed for whole genome SNPs data
genotyped from malaria parasite.</p>
    </div>
    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_snpdata</span><span class="op">(</span></span>
<span>  vcf_file <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  meta_file <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  output_dir <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gof <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gff <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  num_threads <span class="op">=</span> <span class="fl">4L</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>
    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt id="arg-vcf-file">vcf_file<a class="anchor" aria-label="anchor" href="#arg-vcf-file"></a></dt>
<dd><p>the input VCF file (required)</p></dd>
<dt id="arg-meta-file">meta_file<a class="anchor" aria-label="anchor" href="#arg-meta-file"></a></dt>
<dd><p>the metadata file (required)</p></dd>
<dt id="arg-output-dir">output_dir<a class="anchor" aria-label="anchor" href="#arg-output-dir"></a></dt>
<dd><p>the path to the folder where the output files will
be stored (optional)</p></dd>
<dt id="arg-gof">gof<a class="anchor" aria-label="anchor" href="#arg-gof"></a></dt>
<dd><p>the gene ontology annotation file (optional). If not provided, the
default file obtained from the <a href="https://plasmodb.org/plasmo/app/downloads/Current_Release/Pfalciparum3D7/gaf/" class="external-link">PlasmoDB</a> # nolint: line_length_linter
will be used</p></dd>
<dt id="arg-gff">gff<a class="anchor" aria-label="anchor" href="#arg-gff"></a></dt>
<dd><p>the gene annotation file (optional). If not provided, the default
file obtained from the <a href="https://plasmodb.org/plasmo/app/downloads/Current_Release/Pfalciparum3D7/gff/" class="external-link">PlasmoDB</a> # nolint: line_length_linter
will be used</p></dd>
<dt id="arg-num-threads">num_threads<a class="anchor" aria-label="anchor" href="#arg-num-threads"></a></dt>
<dd><p>the number of threads to be used when reading in the data
from the VCF file. default is 4</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
<p>an object of class <code>SNPdata</code> with 5 elements</p><ol><li><p>meta: a <code>data.frame</code> that contains the sample's metadata</p></li>
<li><p>details: a <code>data.frame</code> with SNPs genomic coordinates, the fraction
of missing data per SNP and the names and descriptions of the gene on
which they are located.</p></li>
<li><p>GT: an integer <code>matrix</code> with the genotype data. 0='reference allele',
1='alternate allele', 2='mixed allele', NA='missing allele'</p></li>
<li><p>vcf: the full path to the VCF file from which the data is generated</p></li>
<li><p>index: an integer. this is used as some kind of label that can tell
the how many times the VCF files has been modified.</p></li>
</ol></div>
    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="va">snpdata</span> <span class="op">&lt;-</span> <span class="fu">get_snpdata</span><span class="op">(</span></span></span>
<span class="r-in"><span>    vcf_file   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Input_Data.vcf.gz"</span>,</span></span>
<span class="r-in"><span>                             package <span class="op">=</span> <span class="st">"mpbr"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    meta_file  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"SampleMetadata.RDS"</span>,</span></span>
<span class="r-in"><span>                             package <span class="op">=</span> <span class="st">"mpbr"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    output_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span> <span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>
      <footer><div class="copyright">
  <p></p><p>Developed by Karim Mané, Alfred Amambua Ngwa, David Jeffries.</p>
</div>
<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.9000.</p>
</div>
      </footer></div>
  </body></html>

