<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>mpbr: Summary statistics • mpbr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="mpbr: Summary statistics">
<meta property="og:description" content="mpbr">
<meta property="og:image" content="/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mpbr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.1</span>
      </span>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/mpbr.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Karim-Mane/mpbr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>mpbr: Summary statistics</h1>
                        <small class="dont-index">Source: <a href="https://github.com/Karim-Mane/mpbr/blob/HEAD/vignettes/mpbr.Rmd" class="external-link"><code>vignettes/mpbr.Rmd</code></a></small>
      <div class="hidden name"><code>mpbr.Rmd</code></div>
    </div>
        <div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Karim-Mane.github.io/mpbr/" class="external-link">mpbr</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="construct-the-snpdata-object">Construct the <code>SNPdata</code> object<a class="anchor" aria-label="anchor" href="#construct-the-snpdata-object"></a>
</h2>
<p>The functions in this package require a <code>SNPdata</code> object.
Objects of this class are generated using the <code><a href="../reference/get_snpdata.html">get_snpdata()</a></code>
function with the following arguments:</p>
<ol style="list-style-type: decimal">
<li>
<strong>vcf_file</strong>: the input <a href="https://samtools.github.io/hts-specs/VCFv4.2.pdf" class="external-link">VCF</a> file
(required).<br>
</li>
<li>
<strong>meta_file</strong>: a tab-delimited file with the samples
metadata (required).</li>
<li>
<strong>output_dir</strong>: the path to the folder where the output
files will be stored (optional).</li>
<li>
<strong>gaf</strong>: a tab-delimited file with the gene ontology
annotation (optional). If not provided, the default file that was
downloaded from <a href="https://plasmodb.org/plasmo/app/downloads/Current_Release/Pfalciparum3D7/gaf/" class="external-link uri">https://plasmodb.org/plasmo/app/downloads/Current_Release/Pfalciparum3D7/gaf/</a>
will be used.<br>
</li>
<li>
<strong>gff</strong>: a tab-delimited file with the gene annotation
file (optional). If not provided, the default file that was downloaded
from <a href="https://plasmodb.org/plasmo/app/downloads/Current_Release/Pfalciparum3D7/gff/" class="external-link uri">https://plasmodb.org/plasmo/app/downloads/Current_Release/Pfalciparum3D7/gff/</a>
will be used.</li>
<li>
<strong>num_threads</strong>: the number of threads to use when
importing the input data.</li>
</ol>
<p>Type <code><a href="../reference/get_snpdata.html">?get_snpdata</a></code> for more details.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snpdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_snpdata.html">get_snpdata</a></span><span class="op">(</span>vcf_file   <span class="op">=</span> <span class="va">vcf</span>,</span>
<span>                       meta_file  <span class="op">=</span> <span class="va">meta_file</span>,</span>
<span>                       output_dir <span class="op">=</span> <span class="va">output_dir</span>,</span>
<span>                       gof        <span class="op">=</span> <span class="va">gof</span>,</span>
<span>                       gff        <span class="op">=</span> <span class="va">gff</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="print-the-snpdata-object">print the <code>SNPdata</code> object<a class="anchor" aria-label="anchor" href="#print-the-snpdata-object"></a>
</h2>
<p>Use the <code><a href="https://rdrr.io/r/base/print.html" class="external-link">print()</a></code> function to visualize the structure of
the <code>SNPdata</code> object created from the
<code><a href="../reference/get_snpdata.html">get_snpdata()</a></code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">snpdata</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="calculate-the-snps-minor-allele-frequency">Calculate the SNPs minor allele frequency<a class="anchor" aria-label="anchor" href="#calculate-the-snps-minor-allele-frequency"></a>
</h2>
<p>the <code><a href="../reference/compute_maf.html">compute_maf()</a></code> requires a <code>SNPdata</code>
object, the name of the genotype table from which the MAF should be
calculated (either <strong>“GT”</strong> for the raw genotypes,
<strong>“Phased”</strong> for phased genotypes,
<strong>“Phased_Imputed”</strong> for phased and imputed
genotypes).<br><code>include_het</code> argument determines whether to account for
heterozygous sites during MAF calculation. The function returns a
<code>SNPdata</code> object with the following 2 extra columns in the
<strong>details</strong> table:</p>
<ul>
<li>
<strong>MAF</strong>: the column with the SNPs minor allele
frequencies</li>
<li>
<strong>MAF_allele</strong>: the column that determines which of the
reference (0) or alternate (1) alleles is the minor allele on that
site.</li>
</ul>
<p>type <code>?compute_MAF</code> for more details.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snpdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_maf.html">compute_maf</a></span><span class="op">(</span><span class="va">snpdata</span>,</span>
<span>                       include_het <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                       mat_name    <span class="op">=</span> <span class="st">"GT"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">snpdata</span><span class="op">$</span><span class="va">details</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="filter-the-snpdata-object">Filter the SNPdata object<a class="anchor" aria-label="anchor" href="#filter-the-snpdata-object"></a>
</h2>
<p>Use the <code><a href="../reference/filter_snps_samples.html">filter_snps_samples()</a></code> function to remove the
SNPs and samples that do not satisfy the user criteria. In the example
below, we intend to discard SNPs and samples where the number of missing
genotypes represent 20% or more the total number of observations. SNPs
with MAF &lt; 1% or a quality score &lt; 10 will also be removed.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snpdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_snps_samples.html">filter_snps_samples</a></span><span class="op">(</span><span class="va">snpdata</span>,</span>
<span>                               min_qual            <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                               max_missing_sites   <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>                               max_missing_samples <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>                               maf_cutoff          <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span></code></pre></div>
<p>All loci and samples that do not satisfy the defined conditions will
be dropped from the input <code>SNPdata</code> object. Details about the
filtered object can be viewed using:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">snpdata</span><span class="op">)</span></span></code></pre></div>
<p>The distribution of the MAF in the filtered <code>SNPdata</code>
object looks like below:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">snpdata</span><span class="op">$</span><span class="va">details</span><span class="op">$</span><span class="va">MAF</span>, <span class="fl">100</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/grid.html" class="external-link">grid</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="calculate-the-within-host-genetics-diversity-index-fws">Calculate the within-host genetics diversity index (Fws)<a class="anchor" aria-label="anchor" href="#calculate-the-within-host-genetics-diversity-index-fws"></a>
</h2>
<p>The within-hosts fixation index (Fws) metric measures the genetic
diversity of an individual infection relative to the population level
genetic diversity using the <a href="https://github.com/bahlolab/moimix" class="external-link">moimix</a> R package. The
function creates 2 extra columns in the <strong>meta</strong> table. The
<strong>Fws</strong> column contains the Fws values. They vary between 0
(for polyclonal infection) and 1 (for monoclonal infections). The
<strong>COI</strong> column is created from the Fws column such that
infections with Fws greater than the specified threshold are considered
as monoclonal (1). Otherwise, the infection is considered as polyclonal
(2).</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snpdata</span> <span class="op">&lt;-</span> <span class="fu">calculate_fws</span><span class="op">(</span><span class="va">snpdata</span>, threshold <span class="op">=</span> <span class="fl">0.95</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">snpdata</span><span class="op">$</span><span class="va">meta</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="mixed-genotypes-phasing">Mixed genotypes phasing<a class="anchor" aria-label="anchor" href="#mixed-genotypes-phasing"></a>
</h2>
<p>Mixed genotypes are not accounted for in many population genetics
methods as they introduce a extra layer of complexity. Genotype phasing
is the most common way of dealing with mixed genotype. In this package,
we use the <code><a href="../reference/phase_mixed_genotypes.html">phase_mixed_genotypes()</a></code> to phase the mixed
genotypes based on their allelic depth (AD) and the total allele count
on the specific locus through the following procedure:</p>
<ul>
<li>When there is no read supporting both alleles, the mixed genotype is
replaced by the less common allele in the locus. If both alleles are
equally represented in the population, it is replaced based on a
Bernoulli simulation with a probability defined as <span class="math inline">\(\frac{reference_allele_count}{reference_allele_count
+ alternate_allele_count}\)</span>
</li>
<li>When both allele of the mixed genotype are supported by more than 5
reads and the number of reads supporting either the reference or or
alternate allele is 2 times the other allele, it will be replace by the
allele with the less number of reads. Otherwise the mixed allele is
replaced based on a Bernoulli simulation as describe above.</li>
<li>When there is no read supporting one of the allele and the read
count of the other allele is &gt; 5, the mixed allele is replaced with
the one that is not supported. Other it is replaced based on a Bernoulli
simulation as describe above.</li>
</ul>
<p>The process above is repeated <code>nsim</code> times and a new
genotype file is generated and stored in the output directory. After all
the iterations, the correlation between the MAF before phasing and the
MAF from each phased genotype data is calculated. The phased genotype
with the highest correlation will be retained and stored in the output
<code>SNPdata</code> object. We recommend <code>nsim = 50</code> at
least to maximize the change of getting the highest MAF correlation.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snpdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/phase_mixed_genotypes.html">phase_mixed_genotypes</a></span><span class="op">(</span><span class="va">snpdata</span>, nsim <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">snpdata</span><span class="op">$</span><span class="va">Phased</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="impute-the-missing-data">Impute the missing data<a class="anchor" aria-label="anchor" href="#impute-the-missing-data"></a>
</h2>
<p>Some population genetics methods require genotype data with no
missing allele. In {mpbr}, missing genotypes can be imputed using the
<code><a href="../reference/impute_missing_genotypes.html">impute_missing_genotypes()</a></code> use the
<code>impute_missing_genotypes</code> function to impute the missing
genotypes. The <code>genotype</code> argument determines which genotype
table should be used when imputing the missing data
(<strong>“GT”</strong> to impute from raw data,
<strong>“Phased”</strong> to impute from the phased data).</p>
<p>For every missing allele, we compute the reference and alternate
allele frequencies at the given locus and identify the minor allele
frequency (MAF). The missing allele will be replaced based on a
Bernoulli simulation with a probability that is set at the resulting MAF
value.</p>
<p>This process is repeated <code>nsim</code> times and a new genotype
file is generated and stored in the output directory. After all the
iterations, the correlation between the MAF before imputation and the
MAF from each imputed genotype data is calculated. The imputed genotype
with the highest correlation will be retained and stored in the output
<code>SNPdata</code> object. We recommend <code>nsim = 50</code> at
least to maximize the change of getting the highest MAF correlation.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snpdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/impute_missing_genotypes.html">impute_missing_genotypes</a></span><span class="op">(</span><span class="va">snpdata</span>, genotype <span class="op">=</span> <span class="st">"GT"</span>, nsim <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">snpdata</span><span class="op">$</span><span class="va">Imputed</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="select-data-for-a-given-chromosome">Select data for a given chromosome<a class="anchor" aria-label="anchor" href="#select-data-for-a-given-chromosome"></a>
</h2>
<p>The package also contains functions that can be used to subset the
<code>SNPdata</code> object. To generate a <code>SNPdata</code> object
with data from chromosome 7 only (“Pf3D7_07_v3”) from an existing object
of the same class, use the <code><a href="../reference/select_chrom.html">select_chrom()</a></code> function. Note
that the <strong>meta</strong> table in the output object will not have
the <em>percentage_missing_sites</em>, <em>Fws</em>, <em>COI</em>
columns. This is due to the fact that the existing values in these
columns are calculated from the entire dataset and are not specific the
chosen chromosome.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## subset only chromosome 7 data</span></span>
<span><span class="va">chrom7_snpdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select_chrom.html">select_chrom</a></span><span class="op">(</span><span class="va">snpdata</span>, chrom <span class="op">=</span> <span class="st">"Pf3D7_07_v3"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">chrom7_snpdata</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="drop-a-set-of-snps-from-the-data">drop a set of SNPs from the data<a class="anchor" aria-label="anchor" href="#drop-a-set-of-snps-from-the-data"></a>
</h2>
<p>If you want to remove some loci from the current <code>SNPdata</code>
object, use the <code><a href="../reference/drop_snps.html">drop_snps()</a></code> function and provide it with a
data frame with 2 columns named as: <em>Chrom</em> and <em>Pos”</em>,
representing the genomic coordinates of the loci to be dropped. Another
alternative will be to provide the <code>chrom</code> (chromosome name),
<code>start</code> (the start position of the region to be discarded)
and <code>end</code> (the ens position of the region to be
discarded).</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## drop a set of SNPs based on their genomic coordinates</span></span>
<span><span class="va">snp_to_be_dropped</span> <span class="op">&lt;-</span> <span class="va">chrom7_snpdata</span><span class="op">$</span><span class="va">details</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">Chrom</span>, <span class="va">Pos</span><span class="op">)</span></span>
<span><span class="va">idx</span>               <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">snp_to_be_dropped</span><span class="op">)</span><span class="op">)</span>, <span class="fl">100</span>,</span>
<span>                            replace <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">snp_to_be_dropped</span> <span class="op">&lt;-</span> <span class="va">snp_to_be_dropped</span><span class="op">[</span><span class="va">idx</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># there is a problem in the way the index of the vcf file is set when</span></span>
<span><span class="co"># dropping snps.</span></span>
<span><span class="co"># solution: consider replacing index with an appropriated value</span></span>
<span><span class="va">reduced_snpdata</span>   <span class="op">&lt;-</span> <span class="fu"><a href="../reference/drop_snps.html">drop_snps</a></span><span class="op">(</span><span class="va">chrom7_snpdata</span>, <span class="va">snp_to_be_dropped</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">reduced_snpdata</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># drop a region from a chromosome</span></span>
<span><span class="va">reduced_snpdata</span>   <span class="op">&lt;-</span> <span class="fu"><a href="../reference/drop_snps.html">drop_snps</a></span><span class="op">(</span><span class="va">chrom7_snpdata</span>,</span>
<span>                               chrom <span class="op">=</span> <span class="st">"Pf3D7_07_v3"</span>,</span>
<span>                               start <span class="op">=</span> <span class="fl">110031</span>, end <span class="op">=</span> <span class="fl">154906</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">reduced_snpdata</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="drop-a-set-of-samples-from-the-data">drop a set of samples from the data<a class="anchor" aria-label="anchor" href="#drop-a-set-of-samples-from-the-data"></a>
</h2>
<p>When you want to remove some samples from the <code>SNPdata</code>
object, use the <code><a href="../reference/drop_samples.html">drop_samples()</a></code> function and provide it with
a vector of sample IDs to drop.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">idx</span>                   <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">reduced_snpdata</span><span class="op">$</span><span class="va">meta</span><span class="op">)</span><span class="op">)</span>, <span class="fl">10</span>,</span>
<span>                                replace <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">samples_to_be_dropped</span> <span class="op">&lt;-</span> <span class="va">reduced_snpdata</span><span class="op">$</span><span class="va">meta</span><span class="op">$</span><span class="va">sample</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span></span>
<span><span class="va">reduced_snpdata</span>       <span class="op">&lt;-</span> <span class="fu"><a href="../reference/drop_samples.html">drop_samples</a></span><span class="op">(</span><span class="va">reduced_snpdata</span>, <span class="va">samples_to_be_dropped</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">reduced_snpdata</span><span class="op">)</span></span></code></pre></div>
</div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>
</div>
      <footer><div class="copyright">
  <p></p>
<p>Developed by Karim Mané, Alfred Amambua Ngwa, David Jeffries.</p>
</div>
<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.9000.</p>
</div>
      </footer>
</div>
  </body>
</html>
